### PR TITLE
[DebugInfo] Fix #76058: missing debug location in processSDiv function

### LINKED ISSUE #76058: [DebugInfo] Missing debug location after replacement in processSDiv function
Hi, I found a bug of updating debug locationg in the LLVM transformer **CorrelatedValuePropagation**. After assigning a newly created binary operator Neg, the original debug location of SDI is lost. However, the newly created Neg should preserve the debug location of the SDI. This is the [buggy code](https://github.com/llvm/llvm-project/blob/6ab663be8da2b2f0a3a59d1ab6935100fb832049/llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp#L942). 

The following is the testcase (sdiv_missing_debugloc.ll):
```text
; RUN: opt -passes=correlated-propagation -S < %s | FileCheck %s
; CHECK: %{{[a-zA-Z0-9_]*}} = udiv i8 %x.nonneg, %y, !dbg ![[DBGLOC:[0-9]+]]
; CHECK-NEXT: %{{[a-zA-Z0-9_]*}}.neg = sub i8 0, %rem1, !dbg ![[DBGLOC]]

; ModuleID = 'reduced.ll'
source_filename = "orig.ll"
target datalayout = "e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64"
target triple = "thumbv7m-arm-none-eabi"

; Function Attrs: inaccessiblememonly nocallback nofree nosync nounwind willreturn
declare void @llvm.assume(i1 noundef) #0

define void @test8_neg_neg(i8 %x, i8 %y) !dbg !5 {
  %c0 = icmp sle i8 %x, 0, !dbg !13
  call void @llvm.dbg.value(metadata i1 %c0, metadata !9, metadata !DIExpression()), !dbg !13
  call void @llvm.assume(i1 %c0), !dbg !14
  %c1 = icmp sge i8 %y, 0, !dbg !15
  call void @llvm.dbg.value(metadata i1 %c1, metadata !11, metadata !DIExpression()), !dbg !15
  call void @llvm.assume(i1 %c1), !dbg !16
  %rem = sdiv i8 %x, %y, !dbg !17
  call void @llvm.dbg.value(metadata i8 %rem, metadata !12, metadata !DIExpression()), !dbg !17
  ret void, !dbg !18
}

; Function Attrs: nocallback nofree nosync nounwind readnone speculatable willreturn
declare void @llvm.dbg.value(metadata, metadata, metadata) #1

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}
!llvm.debugify = !{!3, !4}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "debugify", isOptimized: true, runtimeVersion: 0, emissionKind: FullDebug)
!2 = !DIFile(filename: "reduced.ll", directory: "/")
!3 = !{i32 6}
!4 = !{i32 3}
!5 = distinct !DISubprogram(name: "test8_neg_neg", linkageName: "test8_neg_neg", scope: null, file: !2, line: 1, type: !6, scopeLine: 1, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !1, retainedNodes: !8)
!6 = !DISubroutineType(types: !7)
!7 = !{}
!8 = !{!9, !11, !12}
!9 = !DILocalVariable(name: "1", scope: !5, file: !2, line: 1, type: !10)
!10 = !DIBasicType(name: "ty8", size: 8, encoding: DW_ATE_unsigned)
!11 = !DILocalVariable(name: "2", scope: !5, file: !2, line: 3, type: !10)
!12 = !DILocalVariable(name: "3", scope: !5, file: !2, line: 5, type: !10)
!13 = !DILocation(line: 1, column: 1, scope: !5)
!14 = !DILocation(line: 2, column: 1, scope: !5)
!15 = !DILocation(line: 3, column: 1, scope: !5)
!16 = !DILocation(line: 4, column: 1, scope: !5)
!17 = !DILocation(line: 5, column: 1, scope: !5)
!18 = !DILocation(line: 6, column: 1, scope: !5)
```
Apply CorrelatedValuePropagation on it: 
```bash
$ opt -passes=correlated-propagation -S sdiv_missing_debugloc.ll -o opted_sdiv.ll
```
The wrong IR in optimized .ll file is as following:
```
%rem1 = udiv i8 %x.nonneg, %y, !dbg !17
%rem1.neg = sub i8 0, %rem1
```
The second line misses the debug location.

Reproduction: https://godbolt.org/z/T6Kfv66bc

I guess I know how to fix it. How can I do to give the patch?

### PR DESCRIPTION
**Pull request by Apochens**
This PR fixes #76058

### CODE REVIEW

File: llvm/test/Transforms/CorrelatedValuePropagation/sdiv_missing_debugloc.ll [hunk 1]
```diff
@@ -0,0 +1,50 @@
+; RUN: opt -passes=correlated-propagation -S < %s | FileCheck %s
+; CHECK: %{{[a-zA-Z0-9_]*}} = udiv i8 %x.nonneg, %y, !dbg ![[DBGLOC:[0-9]+]]
+; CHECK-NEXT: %{{[a-zA-Z0-9_]*}}.neg = sub i8 0, %rem1, !dbg ![[DBGLOC]]
```
Comment (nikic) [inline llvm/test/Transforms/CorrelatedValuePropagation/sdiv_missing_debugloc.ll:3]: Please move these check lines lower (e.g. directly above the function).

File: llvm/test/Transforms/CorrelatedValuePropagation/sdiv_missing_debugloc.ll [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+; ModuleID = 'reduced.ll'
+source_filename = "orig.ll"
+target datalayout = "e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64"
+target triple = "thumbv7m-arm-none-eabi"
```
Comment (nikic) [inline llvm/test/Transforms/CorrelatedValuePropagation/sdiv_missing_debugloc.ll:8]: Remove these four lines.

File: llvm/test/Transforms/CorrelatedValuePropagation/sdiv_missing_debugloc.ll [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  call void @llvm.dbg.value(metadata i8 %rem, metadata !12, metadata !DIExpression()), !dbg !17
+  ret void, !dbg !18
+}
+
+; Function Attrs: nocallback nofree nosync nounwind readnone speculatable willreturn
+declare void @llvm.dbg.value(metadata, metadata, metadata) #1
+
+!llvm.module.flags = !{!0}
+!llvm.dbg.cu = !{!1}
+!llvm.debugify = !{!3, !4}
```
Comment (nikic) [inline llvm/test/Transforms/CorrelatedValuePropagation/sdiv_missing_debugloc.ll:30]: Can llvm.debugify metadata be dropped?

**Review by nikic — COMMENTED** [review]
Looks fine, some test nits.

File: llvm/test/Transforms/CorrelatedValuePropagation/sdiv_missing_debugloc.ll (hunk 2 above)
Comment (Apochens) [reply llvm/test/Transforms/CorrelatedValuePropagation/sdiv_missing_debugloc.ll:8]: Ok, I'll remove these lines.

File: llvm/test/Transforms/CorrelatedValuePropagation/sdiv_missing_debugloc.ll (hunk 3 above)
Comment (Apochens) [reply llvm/test/Transforms/CorrelatedValuePropagation/sdiv_missing_debugloc.ll:30]: This testcase is actually obtained by modifying another similar testcase [srem_missing_debugloc.ll](https://github.com/llvm/llvm-project/blob/ebd0249fcf672ca164e6cc21379f1c62807b6d56/llvm/test/Transforms/CorrelatedValuePropagation/srem_missing_debugloc.ll). So, I don't know whether the llvm.debugify metatdata can be dropped.

**Comment by Apochens** [comment]
Hi, Nikita! To refine the testcase, I **(1) removed line 5-8**, moving the check lines directly above the function, and **(2) discarded the llvm.debugify metadata** (since it's not necessary). After the refinement, I tested the patch on this testcase, and It does work. 

Could you please review this PR again to check whether there is any other issue? Thanks.

**Review by nikic — APPROVED** [review]

**Pull request merged**